commands/functions.go

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2018 The Doctl Authors All rights reserved.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package commands
+
+import (
+	"github.com/digitalocean/doctl"
+	"github.com/spf13/cobra"
+)
+
+func Functions() *Command {
+	cmd := &Command{
+		Command: &cobra.Command{
+			Use:   "functions",
+			Short: "Work with the functions of your sandbox",
+			Long: `The subcommands of ` + "`" + `doctl sandbox functions` + "`" + ` operate on the deployed (cloud-resident) functions of your sandbox.
+You are able to inspect and list these functions to know what is deployed.  You can also invoke functions to test them.`,
+			Aliases: []string{"fn"},
+		},
+	}
+
+	get := cmdBuilderWithInit(cmd, RunFunctionsGet, "get <functionName>", "Retrieves the deployed copy of a function (code or metadata)",
+		`Use `+"`"+`doctl sandbox functions get`+"`"+` to obtain the code or metadata of a deployed function.
+This allows you to inspect the deployed copy and ascertain whether it corresponds to what
+is in your sandbox area in the local file system.`,
+		Writer, false)
+	AddBoolFlag(get, "url", "r", false, "get function url")
+	AddBoolFlag(get, "code", "", false, "show function code (only works if code is not a zip file)")
+	AddStringFlag(get, "save-env", "E", "", "save environment variables to FILE as key-value pairs")
+	AddStringFlag(get, "save-env-json", "J", "", "save environment variables to FILE as JSON")
+	AddBoolFlag(get, "save", "", false, "save function code to file corresponding to the function name")
+	AddStringFlag(get, "save-as", "", "", "file to save function code to")
+
+	invoke := cmdBuilderWithInit(cmd, RunFunctionsInvoke, "invoke <functionName>", "Invokes a function",
+		`Use `+"`"+`doctl sandbox functions invoke`+"`"+` to invoke a function of your sandbox that has been deployed
+to the cloud.  You can provide inputs and inspect outputs.`,
+		Writer, false)
+	AddBoolFlag(invoke, "web", "", false, "Invoke as a web function, show result as web page")
+	AddStringFlag(invoke, "param", "p", "", "parameter values in KEY VALUE format")
```

Comment:
Trying to do some manual testing and struggling with this. Looking at `nim`, it seems like there is an important difference in how its framework parses flags.

With: 

    nim action:invoke hello --param name world

`name world` are seen as the the input to the `--param` flag.

With:

    ./builds/doctl sandbox functions invoke hello --param=name world

`world` is seen as a positional argument. and returns `Error: (functions.invoke) command contains unsupported arguments`

With:

    ./builds/doctl sandbox functions invoke hello --param="name world"

It is seen as a single value and not a key/value pair and returns `Error: Please provide correct values for flags`

https://github.com/adobe/aio-lib-runtime/blob/becabb33bcd9d95cd3e5da9d8a127a997ccf8fff/src/utils.js#L508

Reply:
Yes.   Embarrassed that I tested `fn invoke` only in the default case (no arguments passed).   The argument passing is broken in this case.

The likely reason is that Adobe I/O uses an atypical parsing pattern for a few of its commands (listed here https://github.com/digitalocean/doctl-sandbox-plugin/blob/a795ffcc72f21e7cbc5ca1fa7df3b8fa912bc2c4/generateCobraWrappers.js#L31).   Oclif actually supports this but I forget the details.  I went to the trouble to add the `multiple` flag as per that table during codegen but then I don't do anything useful with it.  If you look at the table, the only command that we are relying on now in `doctl` is `action:invoke` so hopefully I will not need a completely general fix to fix this.   Will investigate and fix as soon as possible.

Reply:
This is going to be a limited fix that cannot align exactly with what `nim` does unless you have some insight into cobra options for supporting flags with multiple values and multiple repetitions of the same flag.  AFAIK this is not support currently in cobra but I am hardly an expert on cobra.

The first error
```
(functions.invoke) command contains unsupported arguments
```
comes from cobra and the sandbox isn't even invoked.   So, without some help from cobra we can't support `--param name world` (no quotes, multiple values following the flag).   

On the other hand, we could support `--param "name world"` (quoting the pair) by adding a small fixup (working on it now).

We will hit another cobra limitation because the `--param` flag is also supposed to be capable of being repeated to specify more than one parameter (this is allowed by `oclif` and by `nim`.   But, you can specify multiple parameters like this: `--param "name world address universe"` (specifying multiple parameter pairs with a single `--param` flag.

Reply:
I have posted a "fix" to this problem (note that quoting is currently needed, so it is not 100% compatible with syntax accepted by `nim`).   But, at least you can invoke functions with parameters 😃 .   You might be a bit concerned (as am I) at the growth of special cases in the `getFlagArgsArray` function but I'm hoping we can discuss the right way to clean that up and hit it on the next pass.

Reply:
I can confirm the change gets this working as documented. I was just looking at `getFlagArgsArray` and thinking it could use some tests :wink: 

Cobra is using `spf13/pflag` under the hood for flags which has `StringSlice` that supports repeated flags.

https://pkg.go.dev/github.com/spf13/pflag#StringSlice

I'd love to drive usage closer to other doctl commands where possible. We have a pattern of using colon separated (e.g. `key:value`) pairs  (see `doctl compute firewall create` for just one example). Paired with `StringSlice`, you can have either `--param foo:bar,name:world` or `--param foo:bar --param name:world` How's that sound?

Reply:
I did end up writing a separate ticket.